server/create.go

```diff
@@ -0,0 +1,628 @@
+		}
+		layers = append(layers, layer)
+	}
+	return layers, nil
+}
+
+func setLicense(layers []Layer, l string) ([]Layer, error) {
+	blob := strings.NewReader(l)
+	layer, err := NewLayer(blob, "application/vnd.ollama.image.license")
+	if err != nil {
+		return nil, err
+	}
+	layers = append(layers, layer)
+	return layers, nil
+}
+
+func setParameters(layers []Layer, p map[string]any) ([]Layer, error) {
+	for _, layer := range layers {
+		if layer.MediaType != "application/vnd.ollama.image.parameters" {
+			continue
+		}
+
+		digestPath, err := GetBlobsPath(layer.Digest)
+		if err != nil {
+			return nil, err
+		}
+
+		fn, err := os.Open(digestPath)
+		if err != nil {
+			return nil, err
+		}
+		defer fn.Close()
+
+		var existing map[string]any
+		if err := json.NewDecoder(fn).Decode(&existing); err != nil {
+			return nil, err
+		}
+
+		for k, v := range existing {
+			if _, exists := p[k]; exists {
+				continue
+			}
+			p[k] = v
+		}
+	}
+
+	if len(p) == 0 {
+		return layers, nil
+	}
+
+	layers = removeLayer(layers, "application/vnd.ollama.image.parameters")
+
+	var b bytes.Buffer
+	if err := json.NewEncoder(&b).Encode(p); err != nil {
+		return nil, err
+	}
+	layer, err := NewLayer(&b, "application/vnd.ollama.image.parameters")
+	if err != nil {
+		return nil, err
+	}
+	layers = append(layers, layer)
+	return layers, nil
+}
+
+func setMessages(layers []Layer, m []api.Message) ([]Layer, error) {
+	/*
+		// this merges the old messages with the new messages
+		for _, layer := range layers {
+			if layer.MediaType != "application/vnd.ollama.image.messages" {
+				continue
+			}
+
+			digestPath, err := GetBlobsPath(layer.Digest)
+			if err != nil {
+				return nil, err
+			}
+
+			fn, err := os.Open(digestPath)
+			if err != nil {
+				return nil, err
+			}
+			defer fn.Close()
+
+			var existing []api.Message
+			if err := json.NewDecoder(fn).Decode(&existing); err != nil {
+				return nil, err
+			}
+			m = append(existing, m...)
+		}
+	*/
+
+	// this leaves the old messages intact if no new messages were specified
+	// which may not be the correct behaviour
+	if len(m) == 0 {
+		return layers, nil
+	}
+
+	layers = removeLayer(layers, "application/vnd.ollama.image.messages")
+	var b bytes.Buffer
+	if err := json.NewEncoder(&b).Encode(m); err != nil {
+		return nil, err
+	}
+	layer, err := NewLayer(&b, "application/vnd.ollama.image.messages")
+	if err != nil {
+		return nil, err
+	}
+	layers = append(layers, layer)
+	return layers, nil
+}
+
+func getConfigLayer(layers []Layer, config ConfigV2) (*Layer, error) {
+	digests := make([]string, len(layers))
+	for i, layer := range layers {
+		digests[i] = layer.Digest
+	}
+	config.RootFS.DiffIDs = digests
+
+	var b bytes.Buffer
+	if err := json.NewEncoder(&b).Encode(config); err != nil {
+		return nil, err
+	}
+	layer, err := NewLayer(&b, "application/vnd.docker.container.image.v1+json")
+	if err != nil {
+		return nil, err
+	}
+	return &layer, nil
+}
+
+func createLink(src, dst string) error {
```

Comment:
Instead of copying each of the files to a new temporary directory, we instead try to symlink the blob to the file name specified in the create request. If we can't make the symlink we'll copy the file instead, but it adds a lot more time.

Reply:
Is an alternative here to modify the convert code to be aware of the sha256 mappings? or perhaps a simple in-memory fs abstraction that maps from `/api/create` files to blobs? Given this is internal it can be in a follow up